### ISSUE TITLE
Test failure: Regressions/coreclr/GitHub_45929/test45929/test45929.cmd

### ISSUE THREAD

**Issue by v-wenyuxu** [dotnet/runtime#104878]
**Failed in:** [runtime-coreclr gcstress-extra 20240714.1](https://dev.azure.com/dnceng-public/public/_build/results?buildId=740628&view=ms.vss-test-web.build-test-results-tab&runId=18691034&resultId=108103&paneView=debug)

**Failed tests:**
```
coreclr linux arm Checked heapverify1 @ (Debian.12.Arm32.Open)Ubuntu.2004.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:debian-12-helix-arm32v7
    - Regressions/coreclr/GitHub_45929/test45929/test45929.cmd
```

**Error message:**
```
 waitpid() returned successfully (wstatus 00000000) WEXITSTATUS 0 WTERMSIG 0
/root/helix/work/workitem/e/Regressions/Regressions/../coreclr/GitHub_45929/test45929/test45929.sh: line 441:  1525 Segmentation fault      (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"

Return code:      1
Raw output file:      /root/helix/work/workitem/uploads/coreclr/GitHub_45929/test45929/output.txt
Raw output:
BEGIN EXECUTION
/root/helix/work/correlation/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false -p System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization=true test45929.dll ''
The test started.
Progress:
7/14/2024 10:22:09 PM : 000.0% : 93
7/14/2024 10:22:10 PM : 000.0% : 514
7/14/2024 10:22:10 PM : 000.0% : 519
7/14/2024 10:22:10 PM : 000.0% : 536
7/14/2024 10:22:10 PM : 000.0% : 564
7/14/2024 10:22:10 PM : 000.1% : 605
7/14/2024 10:22:10 PM : 000.1% : 669
7/14/2024 10:22:10 PM : 000.1% : 755
7/14/2024 10:22:10 PM : 000.2% : 881
7/14/2024 10:22:10 PM : 000.3% : 1099
7/14/2024 10:22:11 PM : 000.4% : 1426
7/14/2024 10:22:11 PM : 000.6% : 1927
7/14/2024 10:22:12 PM : 000.9% : 2660
7/14/2024 10:22:12 PM : 001.0% : 2992
7/14/2024 10:22:15 PM : 002.0% : 5339
7/14/2024 10:22:17 PM : 003.0% : 7691
7/14/2024 10:22:19 PM : 004.0% : 10040
7/14/2024 10:22:22 PM : 005.0% : 12396
7/14/2024 10:22:24 PM : 006.0% : 14749
7/14/2024 10:22:26 PM : 007.0% : 17084
7/14/2024 10:22:29 PM : 008.0% : 19464
[createdump] Gathering state for process 1525 corerun
[createdump] Crashing thread 05f5 signal 11 (000b)
[createdump] Writing crash report to file /home/helixbot/dotnetbuild/dumps/coredump.1525.dmp.crashreport.json
[createdump] Crash report successfully written
[createdump] Writing minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.1525.dmp
[createdump] Written 96194560 bytes (23485 pages) to core file
[createdump] Target process is alive
[createdump] Dump successfully written in 181ms
Expected: 100
Actual: 139
END EXECUTION - FAILED
Test failed. Trying to see if dump file was created in /home/helixbot/dotnetbuild/dumps since 7/14/2024 10:22:09 PM
Processing /home/helixbot/dotnetbuild/dumps/coredump.1525.dmp.crashreport.json
Printing stacktrace from '/home/helixbot/dotnetbuild/dumps/coredump.1525.dmp.crashreport.json'
Invoking llvm-symbolizer --pretty-print
Stack trace:
----------------------------------
Thread Id: 0x5f5
      Child SP               IP Call Site
 0xf7c5ab20 0xf7957614 libc.so.6!
 0xf7c5ab28 0xf79c19c0 libc.so.6!__wait4_time64
 0xf7c5ab90 0xf78556dc libcoreclr.so!PROCCreateCrashDump(std::vector<char const*, std::allocator<char const*> >&, char*, int, bool) at /__w/1/s/src/coreclr/pal/src/thread/process.cpp:2308:22
 0xf7c5abd8 0xf78564dc libcoreclr.so!PROCCreateCrashDumpIfEnabled at /__w/1/s/src/coreclr/pal/src/thread/process.cpp:2526:9
 0xf7c5ac28 0xf780af7e libcoreclr.so!invoke_previous_action(sigaction*, int, siginfo_t*, void*, bool) at /__w/1/s/src/coreclr/pal/src/exception/signal.cpp:422:13
 0xf7c5ac60 0xf780a500 libcoreclr.so!sigsegv_handler(int, siginfo_t*, void*) at /__w/1/s/src/coreclr/pal/src/exception/signal.cpp:0:5
 0xf7c5ac90 0xf79671b0 libc.so.6!
 0xffe15c10 0xf76b3be6 libcoreclr.so!Object::RawGetMethodTable() const at /__w/1/s/src/coreclr/vm/object.h:141:16
 (inlined by) WKS::CObjectHeader::GetMethodTable() const at /__w/1/s/src/coreclr/gc/gc.cpp:4742:44
 (inlined by) WKS::GCHeap::ValidateObjectMember(Object*) at /__w/1/s/src/coreclr/gc/gc.cpp:48081:5
 0xffe15c40 0xf74bb6dc libcoreclr.so!Object::ValidateInner(int, int, int) at /__w/1/s/src/coreclr/vm/object.cpp:588:13
 0xffe15cb8 0xf74bb0ea libcoreclr.so!GetThreadNULLOk() at /__w/1/s/src/coreclr/vm/threads.inl:37:12
 (inlined by) Thread::IsObjRefValid(OBJECTREF const*) at /__w/1/s/src/coreclr/vm/threads.h:3243:34
 (inlined by) OBJECTREF::operator=(OBJECTREF const&) at /__w/1/s/src/coreclr/vm/object.cpp:1335:5
 0xffe15cd8 0xf744e67e libcoreclr.so!StackTraceInfo::EnsureKeepAliveArray(REF<PtrArray>*, unsigned int)
```

**Stack trace:**
```
   at Xunit.Assert.True(Nullable`1 condition, String userMessage) in /_/src/Microsoft.DotNet.XUnitAssert/src/BooleanAsserts.cs:line 146
   at TestLibrary.OutOfProcessTest.RunOutOfProcessTest(String assemblyPath, String testPathPrefix)
   at Program.<<Main>$>g__TestExecutor58|0_59(StreamWriter tempLogSw, StreamWriter statsCsvSw, <>c__DisplayClass0_0&)
```

Queued | Pipeline | Pipeline_Configuration | OS | Arch | Test | Outcome | Properties | BuildReason
-- | -- | -- | -- | -- | -- | -- | -- | --
2024-07-14T07:20:34.843Z | runtime-coreclr crossgen2-composite gcstress | Checked-gcstress0xc_disabler2r_heapverify1 | osx.1200.arm64.open | arm64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"Project":"public","System.PhaseAttempt":"1","DefinitionId":"141","System.StageName":"Build","System.StageAttempt":"1","System.PhaseName":"run_test_p1_Composite_osx_arm64_checked","Reason":"Schedule","operatingSystem":"OSX.1200.ARM64.Open","DefinitionName":"runtime-coreclr   crossgen2-composite gcstress","architecture":"arm64","System.JobName":"__default","System.JobId":"4b5575ba-4f05-5198-bc88-a075c735caa5","CollectionUri":"https://dev.azure.com/dnceng-public/","BuildNumber":"20240713.1","BuildId":"740354","AzurePipelinesTestRunId":"18682728","configuration":"Checked-gcstress0xc_disabler2r_heapverify1","System.JobAttempt":"1"} | Schedule
2024-07-14T07:20:33.413Z | runtime-coreclr crossgen2-composite gcstress | Checked-gcstress0xc_jitminopts_heapverify1 | osx.1200.arm64.open | arm64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"AzurePipelinesTestRunId":"18682726","DefinitionName":"runtime-coreclr   crossgen2-composite   gcstress","System.JobAttempt":"1","BuildNumber":"20240713.1","Project":"public","System.PhaseName":"run_test_p1_Composite_osx_arm64_checked","System.JobId":"4b5575ba-4f05-5198-bc88-a075c735caa5","System.StageAttempt":"1","System.StageName":"Build","operatingSystem":"OSX.1200.ARM64.Open","configuration":"Checked-gcstress0xc_jitminopts_heapverify1","System.PhaseAttempt":"1","System.JobName":"__default","BuildId":"740354","DefinitionId":"141","architecture":"arm64","CollectionUri":"https://dev.azure.com/dnceng-public/","Reason":"Schedule"} | Schedule
2024-07-14T07:20:23.897Z | runtime-coreclr crossgen2-composite gcstress | Checked-heapverify1 | osx.1200.arm64.open | arm64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"System.PhaseAttempt":"1","DefinitionId":"141","Project":"public","System.JobName":"__default","CollectionUri":"https://dev.azure.com/dnceng-public/","Reason":"Schedule","System.JobId":"4b5575ba-4f05-5198-bc88-a075c735caa5","BuildNumber":"20240713.1","BuildId":"740354","configuration":"Checked-heapverify1","architecture":"arm64","System.PhaseName":"run_test_p1_Composite_osx_arm64_checked","System.StageAttempt":"1","System.StageName":"Build","AzurePipelinesTestRunId":"18682714","System.JobAttempt":"1","DefinitionName":"runtime-coreclr   crossgen2-composite   gcstress","operatingSystem":"OSX.1200.ARM64.Open"} | Schedule
2024-06-24T05:28:43.423Z | runtime-coreclr jitstress | Checked-jitstress1_tiered | osx.1200.arm64.open | arm64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"System.PhaseAttempt":"1","AzurePipelinesTestRunId":"17975208","System.JobId":"765fd5e9-e47f-5cb0-4cbf-499eaaaee93e","configuration":"Checked-jitstress1_tiered","System.StageAttempt":"1","System.JobName":"__default","architecture":"arm64","BuildId":"717591","Project":"public","System.PhaseName":"run_test_p1__osx_arm64_checked","DefinitionName":"runtime-coreclr   jitstress","BuildNumber":"20240623.1","operatingSystem":"OSX.1200.ARM64.Open","CollectionUri":"https://dev.azure.com/dnceng-public/","System.StageName":"Build","DefinitionId":"109","System.JobAttempt":"1","Reason":"Schedule"} | Schedule
2024-06-24T05:18:01.762Z | runtime-coreclr jitstress | Checked-jitstress1_tiered | osx.1200.amd64.open | x64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"System.PhaseName":"run_test_p1__osx_x64_checked","System.StageName":"Build","DefinitionName":"runtime-coreclr   jitstress","System.JobId":"bf644a0b-2dbe-5a03-d8c6-a86fdd97c823","System.JobAttempt":"1","configuration":"Checked-jitstress1_tiered","CollectionUri":"https://dev.azure.com/dnceng-public/","BuildId":"717591","System.PhaseAttempt":"1","Reason":"Schedule","operatingSystem":"OSX.1200.Amd64.Open","Project":"public","architecture":"x64","DefinitionId":"109","AzurePipelinesTestRunId":"17975056","BuildNumber":"20240623.1","System.JobName":"__default","System.StageAttempt":"1"} | Schedule
2024-06-24T03:23:46.723Z | runtime-coreclr r2r-extra | Checked-jitstress1_tiered | osx.1200.arm64.open | arm64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"System.JobName":"__default","System.StageAttempt":"1","DefinitionId":"114","AzurePipelinesTestRunId":"17974536","System.PhaseName":"run_test_p1_R2R_osx_arm64_checked","DefinitionName":"runtime-coreclr   r2r-extra","CollectionUri":"https://dev.azure.com/dnceng-public/","System.StageName":"Build","System.JobId":"99f15f99-448d-5593-a271-dbdf553789a6","System.PhaseAttempt":"1","architecture":"arm64","System.JobAttempt":"1","configuration":"Checked-jitstress1_tiered","operatingSystem":"OSX.1200.ARM64.Open","Project":"public","BuildNumber":"20240623.1","BuildId":"717577","Reason":"Schedule"} | Schedule
2024-06-24T03:23:43.738Z | runtime-coreclr r2r-extra | Checked-jitstress2_tiered | osx.1200.arm64.open | arm64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"DefinitionName":"runtime-coreclr   r2r-extra","System.StageAttempt":"1","BuildId":"717577","CollectionUri":"https://dev.azure.com/dnceng-public/","System.PhaseAttempt":"1","DefinitionId":"114","AzurePipelinesTestRunId":"17974534","System.PhaseName":"run_test_p1_R2R_osx_arm64_checked","configuration":"Checked-jitstress2_tiered","System.JobAttempt":"1","Reason":"Schedule","System.JobName":"__default","System.JobId":"99f15f99-448d-5593-a271-dbdf553789a6","BuildNumber":"20240623.1","System.StageName":"Build","Project":"public","operatingSystem":"OSX.1200.ARM64.Open","architecture":"arm64"} | Schedule
2024-06-23T05:21:09.233Z | runtime-coreclr jitstress | Checked-jitstress1_tiered | osx.1200.arm64.open | arm64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"operatingSystem":"OSX.1200.ARM64.Open","System.StageName":"Build","configuration":"Checked-jitstress1_tiered","architecture":"arm64","System.PhaseAttempt":"1","System.JobName":"__default","Reason":"Schedule","AzurePipelinesTestRunId":"17958224","System.JobAttempt":"1","BuildNumber":"20240622.1","Project":"public","System.StageAttempt":"1","DefinitionName":"runtime-coreclr   jitstress","CollectionUri":"https://dev.azure.com/dnceng-public/","DefinitionId":"109","System.JobId":"765fd5e9-e47f-5cb0-4cbf-499eaaaee93e","BuildId":"717159","System.PhaseName":"run_test_p1__osx_arm64_checked"} | Schedule
2024-06-22T05:22:55.752Z | runtime-coreclr jitstress | Checked-jitstress1_tiered | osx.1200.arm64.open | arm64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"operatingSystem":"OSX.1200.ARM64.Open","System.StageName":"Build","architecture":"arm64","System.JobId":"765fd5e9-e47f-5cb0-4cbf-499eaaaee93e","configuration":"Checked-jitstress1_tiered","System.StageAttempt":"1","System.PhaseAttempt":"1","BuildId":"716601","DefinitionName":"runtime-coreclr   jitstress","DefinitionId":"109","System.JobAttempt":"1","BuildNumber":"20240621.1","Reason":"Schedule","Project":"public","CollectionUri":"https://dev.azure.com/dnceng-public/","AzurePipelinesTestRunId":"17939412","System.PhaseName":"run_test_p1__osx_arm64_checked","System.JobName":"__default"} | Schedule
2024-06-22T05:22:35.329Z | runtime-coreclr jitstress | Checked-jitstress1_tiered | osx.1200.amd64.open | x64 | Regressions/coreclr/GitHub_45929/test45929/test45929.cmd | Failed | {"AzurePipelinesTestRunId":"17939402","BuildNumber":"20240621.1","System.JobId":"bf644a0b-2dbe-5a03-d8c6-a86fdd97c823","DefinitionName":"runtime-coreclr   jitstress","configuration":"Checked-jitstress1_tiered","architecture":"x64","System.JobAttempt":"1","System.JobName":"__default","System.StageName":"Build","System.PhaseAttempt":"1","System.PhaseName":"run_test_p1__osx_x64_checked","Project":"public","BuildId":"716601","DefinitionId":"109","Reason":"Schedule","System.StageAttempt":"1","operatingSystem":"OSX.1200.Amd64.Open","CollectionUri":"https://dev.azure.com/dnceng-public/"} | Schedule

**Comment by jakobbotsch** [comment]
Stack trace of crash looks related to stack trace handling:
```
Thread Id: 0x5f5
      Child SP               IP Call Site
 0xf7c5ab20 0xf7957614 libc.so.6!
 0xf7c5ab28 0xf79c19c0 libc.so.6!__wait4_time64
 0xf7c5ab90 0xf78556dc libcoreclr.so!PROCCreateCrashDump(std::vector<char const*, std::allocator<char const*> >&, char*, int, bool) at /__w/1/s/src/coreclr/pal/src/thread/process.cpp:2308:22
 0xf7c5abd8 0xf78564dc libcoreclr.so!PROCCreateCrashDumpIfEnabled at /__w/1/s/src/coreclr/pal/src/thread/process.cpp:2526:9
 0xf7c5ac28 0xf780af7e libcoreclr.so!invoke_previous_action(sigaction*, int, siginfo_t*, void*, bool) at /__w/1/s/src/coreclr/pal/src/exception/signal.cpp:422:13
 0xf7c5ac60 0xf780a500 libcoreclr.so!sigsegv_handler(int, siginfo_t*, void*) at /__w/1/s/src/coreclr/pal/src/exception/signal.cpp:0:5
 0xf7c5ac90 0xf79671b0 libc.so.6!
 0xffe15c10 0xf76b3be6 libcoreclr.so!Object::RawGetMethodTable() const at /__w/1/s/src/coreclr/vm/object.h:141:16
 (inlined by) WKS::CObjectHeader::GetMethodTable() const at /__w/1/s/src/coreclr/gc/gc.cpp:4742:44
 (inlined by) WKS::GCHeap::ValidateObjectMember(Object*) at /__w/1/s/src/coreclr/gc/gc.cpp:48081:5
 0xffe15c40 0xf74bb6dc libcoreclr.so!Object::ValidateInner(int, int, int) at /__w/1/s/src/coreclr/vm/object.cpp:588:13
 0xffe15cb8 0xf74bb0ea libcoreclr.so!GetThreadNULLOk() at /__w/1/s/src/coreclr/vm/threads.inl:37:12
 (inlined by) Thread::IsObjRefValid(OBJECTREF const*) at /__w/1/s/src/coreclr/vm/threads.h:3243:34
 (inlined by) OBJECTREF::operator=(OBJECTREF const&) at /__w/1/s/src/coreclr/vm/object.cpp:1335:5
 0xffe15cd8 0xf744e67e libcoreclr.so!StackTraceInfo::EnsureKeepAliveArray(REF<PtrArray>*, unsigned int) at /__w/1/s/src/coreclr/vm/vars.hpp:0:7
 0xffe15d20 0xf744ec7e libcoreclr.so!StackTraceInfo::AppendElement(OBJECTHANDLE__*, unsigned int, unsigned int, MethodDesc*, CrawlFrame*) at /__w/1/s/src/coreclr/vm/excep.cpp:0:13
 0xffe15dd8 0xf7637b3c libcoreclr.so!AppendExceptionStackFrame at /__w/1/s/src/coreclr/vm/exceptionhandling.cpp:0:13
```

@janvorli Do you think this is most likely a codegen issue?

**Comment by jakobbotsch** [comment]
The failure does not look isolated to arm32. It looks to be reproing on most platforms in gcstress-extra.

**Comment by jakobbotsch** [comment]
A bisection suggests this was introduced or exposed by f99194cb7f227564e9acb4c2a3e22f1b0d3cbe54, cc @amanasifkhalid

**Comment by janvorli** [comment]
Ah, at least the initial link in the issue works and I can find failures in the history. It looks like a GC hole, I'll try to repro it locally.

**Comment by janvorli** [comment]
I can repro it, the issue is not coming from the @amanasifkhalid's change but rather from the commit right before that where I have removed a global spinlock for exception stack traces and reworked the stack trace updating significantly.
This problem wasn't caught by the PR, I've ran the GC stress tests 3 and C - and also ran them countless times locally. The HeapVerify in the gcstress-extra is what reveals the problem.
